File: backend/app/schemas/collector.py
"""采集器管理 API 的 Pydantic schemas。"""

from datetime import datetime

from pydantic import Field

from app.schemas.base import CamelModel


class CollectorTaskRunRequest(CamelModel):
    """触发采集器任务的可选运行参数。"""

    preferred_source: str | None = Field(None, max_length=50)
    symbols: list[str] | None = Field(None, max_length=100)
    period: str | None = Field(None, max_length=20)
    start_date: str | None = Field(None, max_length=20)
    end_date: str | None = Field(None, max_length=20)
    sector_type: str | None = Field(None, max_length=20)
    indicators: list[str] | None = Field(None, max_length=20)
    report_types: list[str] | None = Field(None, max_length=20)
    report_date: str | None = Field(None, max_length=20)
    trade_date: str | None = Field(None, max_length=20)


class CollectorTaskChannelItem(CamelModel):
    """任务可用的单个渠道。"""

    source: str
    name: str
    is_enabled: bool


class CollectorTaskChannelsResponse(CamelModel):
    """采集器任务的可用渠道及解析出的默认渠道。"""

    task_name: str
    data_type: str
    channels: list[CollectorTaskChannelItem]
    resolved_source: str | None


class CollectorTaskCatalogItem(CamelModel):
    """任务目录项（由注册表 TASK_SPECS 派生）。"""

    name: str
    label: str
    data_type: str
    sources: list[str]
    config_params: list[str]
    run_params: list[str]


class CollectorTaskCatalogResponse(CamelModel):
    """任务目录：管理端 UI 触发列表的唯一数据源。"""

    items: list[CollectorTaskCatalogItem]


class CollectorRunResponse(CamelModel):
    """接受采集触发请求后返回的响应。"""

    task_name: str
    status: str = "accepted"
    log_id: int | None = None
    celery_task_id: str | None = None


class CollectorLogResponse(CamelModel):
    """单条采集器执行日志。

    ORM 列名为 ``meta``，wire 输出 ``metadata``：validation_alias 只影响
    读取来源，序列化仍走 to_camel 的字段名。
    """

    id: int
    task_name: str
    source: str | None
    status: str
    celery_task_id: str | None = None
    started_at: datetime | None
    finished_at: datetime | None
    records_count: int
    error_msg: str | None
    metadata: dict | None = Field(default=None, validation_alias="meta")


class CollectorDeadLetterResponse(CamelModel):
    """单条采集器死信记录。"""

    id: int
    task_name: str
    source: str | None
    payload: dict
    celery_task_id: str | None
    collector_log_id: int | None
    error_msg: str | None
    retry_count: int
    created_at: datetime

File: shared/types/collector.ts
/** 采集渠道配置 masked 视图（对应 admin/collector 渠道管理）。 */
export interface ApiCollectorChannelConfigResponse {
  id: number
  source: string
  name: string
  baseUrl: string | null
  apiKeyMasked: string | null
  isEnabled: boolean
  supportedDataTypes: string[]
  extra: Record<string, unknown>
  proxyConfigId: number | null
  createdAt: string
  updatedAt: string
}

export interface ApiCollectorChannelConfigCreateRequest {
  source: string
  name: string
  baseUrl?: string
  apiKey?: string
  isEnabled?: boolean
  supportedDataTypes?: string[]
  extra?: Record<string, unknown>
  proxyConfigId?: number | null
}

export interface ApiCollectorChannelConfigUpdateRequest {
  name?: string
  baseUrl?: string
  apiKey?: string
  isEnabled?: boolean
  supportedDataTypes?: string[]
  extra?: Record<string, unknown>
  proxyConfigId?: number | null
}

export interface ApiProxyConfigResponse {
  id: number
  name: string
  protocol: 'http' | 'socks5'
  host: string
  port: number
  username: string | null
  passwordMasked: string | null
  isEnabled: boolean
  createdAt: string
  updatedAt: string
}

export interface ApiProxyConfigCreateRequest {
  name: string
  protocol: 'http' | 'socks5'
  host: string
  port: number
  username?: string
  password?: string
  isEnabled?: boolean
}

/** password 留空/缺省表示不修改已存储的密码。 */
export interface ApiProxyConfigUpdateRequest {
  name?: string
  protocol?: 'http' | 'socks5'
  host?: string
  port?: number
  username?: string | null
  password?: string
  isEnabled?: boolean
}

export interface ApiProxyConfigTestResponse {
  ok: boolean
  statusCode: number | null
  latencyMs: number
  error: string | null
}

export interface ApiCollectorTaskChannelItem {
  source: string
  name: string
  isEnabled: boolean
}

export interface ApiCollectorTaskChannelsResponse {
  taskName: string
  dataType: string
  channels: ApiCollectorTaskChannelItem[]
  resolvedSource: string | null
}

export interface ApiCollectorTaskRunRequest {
  preferredSource?: string | null
  symbols?: string[] | null
  period?: string | null
  startDate?: string | null
  endDate?: string | null
  sectorType?: string | null
  indicators?: string[] | null
  reportTypes?: string[] | null
  reportDate?: string | null
  tradeDate?: string | null
}

export interface ApiCollectorTaskCatalogItem {
  name: string
  label: string
  description: string
  dataType: string
  sources: string[]
  configParams: string[]
  runParams: string[]
  defaults?: Record<string, unknown> | null
}

export interface ApiCollectorTaskCatalogResponse {
  items: ApiCollectorTaskCatalogItem[]
}

export interface ApiCollectorRunResponse {
  taskName: string
  status: string
  logId?: number | null
  celeryTaskId?: string | null
}

export interface ApiCollectorLogResponse {
  id: number
  taskName: string
  source: string | null
  status: string
  startedAt: string | null
  finishedAt: string | null
  recordsCount: number
  errorMsg: string | null
  message?: string | null
  metadata: Record<string, unknown> | null
}

export interface ApiCollectorLogSummaryResponse {
  date: string
  successCount: number
  partialCount: number
  failedCount: number
  skippedCount: number
  runningCount: number
  pendingCount: number
}

// ---------------------------------------------------------------------------
// 采集健康监测（F-MON 一期）
// ---------------------------------------------------------------------------

export type CollectorHealthStatus =
  | 'healthy'
  | 'degraded'
  | 'critical'
  | 'silent'
  | 'paused'
  | 'unconfigured'

export type CollectorHealthRole = 'primary' | 'backup' | 'single'

export type CollectorErrorCause =
  | 'waf'
  | 'network'
  | 'parse'
  | 'auth'
  | 'timeout'
  | 'not_ready'
  | 'other'

export interface ApiCollectorHealthCounts {
  healthy: number
  degraded: number
  critical: number
  silent: number
  paused: number
  unconfigured: number
}

export interface ApiCollectorHealthDomainSummary {
  domain: string
  total: number
  healthy: number
  degraded: number
  critical: number
  silent: number
}

export interface ApiCollectorHealthOverview {
  healthScore: number
  total: number
  counts: ApiCollectorHealthCounts
  successRate24h: number | null
  domains: ApiCollectorHealthDomainSummary[]
  checkedAt: string | null
  staleAfter: string | null
}

export interface ApiCollectorHealthTaskItem {
  taskType: string
  source: string
  status: CollectorHealthStatus
  role: CollectorHealthRole
  domain: string
  successRate24h: number | null
  successRate7d: number | null
  consecutiveFailures: number
  windowsWithoutSuccess: number
  lastSuccessAt: string | null
  lastErrorSummary: string | null
  lastErrorCause: CollectorErrorCause | null
  /** 判定依据（why，按序）——解释该状态如何得出 */
  reasons: string[]
  isHighFrequency: boolean
  lastRecordsCount: number | null
  lastRecordsDate: string | null
  stateChangedAt: string
  checkedAt: string
  schedule: string | null
  isActive: boolean | null
}

export interface ApiCollectorChannelHealthItem {
  source: string
  domainCount: number
  instanceCount: number
  successRate7d: number | null
  faultCount: number
  causes: Record<string, number>
}

export interface ApiCollectorScheduleCheckItem {
  taskType: string
  source: string
  domain: string
  role: CollectorHealthRole
  isActive: boolean
  hasTaskRow: boolean
  schedule: string | null
  windowTotal: number
  successWindows: number
  skippedWindows: number
  failedWindows: number
  missingWindows: number
  exempted: boolean
  lastErrorSummary: string | null
  lastErrorCause: CollectorErrorCause | null
}

export interface ApiCollectorScheduleCheckResponse {
  date: string
  isTradeDay: boolean
  items: ApiCollectorScheduleCheckItem[]
}

export interface ApiCollectorRunHealthCheckResponse {
  checkedAt: string
  total: number
  failed: number
  orphaned: number
  statusCounts: Record<string, number>
}

export interface ApiCollectorClearSnapshotsResponse {
  deleted: number
}
